java/org/contikios/cooja/Simulation.java

```diff
@@ -352,30 +405,45 @@ public void stopSimulation() {
    * @param rv Return value from startSimulation, should be null unless > 0
    */
   public void stopSimulation(boolean block, Integer rv) {
-    if (!isRunning()) {
+    if (!isRunning() || isShutdown) {
       return;
     }
     assert rv == null || rv > 0 : "Pass in rv = null or rv > 0";
     if (rv != null) {
       returnValue = rv;
     }
-    stopSimulation = true;
+
+    commandQueue.add(STOP);
 
     if (block) {
-      if (Thread.currentThread() == simulationThread) {
-        return;
-      }
+      waitFor(false, 250);
+    }
+  }
 
-      /* Wait until simulation stops */
-      try {
-        Thread simThread = simulationThread;
-        if (simThread != null) {
-          simThread.join(100);
+  private void waitFor(boolean isRunning, long maxTime) {
+    if (Thread.currentThread() == simulationThread) {
+      return;
+    }
+    try {
+      long startTime = System.currentTimeMillis();
```

Comment:
This shouldn't throw exceptions, hoist outside the try-block.

Reply:
Good point. I will update this.